cache/lru/lru.go

```diff
@@ -155,24 +155,53 @@ func (c *Cache) Clear() {
 
 // Range call f on every key
 func (c *Cache) Range(f func(key Key, value interface{}) bool) {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
+	var (
+		preCache []entry
+	)
+	c.mutex.RLock()
 	for _, e := range c.cache {
-		kv := e.Value.(*entry)
-		if !f(kv.key, kv.value) {
+		if e == nil {
+			continue
+		}
+		if kv := e.Value.(*entry); kv != nil {
+			preCache = append(preCache, *kv)
+		}
+	}
+	c.mutex.RUnlock()
+	for _, v := range preCache {
+		if !f(v.key, v.value) {
+			// TODO: should not break the loop once failed
 			break
 		}
 	}
 }
 
 // RangeEvictOnError removes the key if call failed
 func (c *Cache) RangeEvictOnError(f func(key Key, value interface{}) error) {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
+	var (
```

Comment:
don't need change here, deleting is safe, from https://golang.org/ref/spec#For_statements:

> If a map entry that has not yet been reached is removed during iteration, the corresponding iteration value will not be produced.

we also have a test that use `RangeEvictOnError()` to delete 5000 items out of 10000, the test pass without any problem